### PR TITLE
REST API: Update and delete metrics

### PR DESCRIPTION
**Pull request by bttf**
This adds `PUT` and `DELETE` endpoints to our `/metrics` branch of the REST API, allowing users to update and delete metrics programmatically!

For `PUT` operations, we basically follow the same input structure as `POST`, with the omission of a few arguments that are not applicable (e.g. datasource, organization). We don't mark any top-level fields as required, but still mark fields within nested objects as required.

Changes:
- Add new routes `PUT` and `DELETE`
- Refactor post validation method to support both post and put (creating & updating)
- Add new putMetricPayloadtoMetricInterface method
  - Outputs a partial MetricInterface, as intended for an update operation
- Add suite of tests based on existing tests for post operation

# Screenshots

### Updating Metric
Screenshot 2023-06-26 at 3 47 16 PM

Screenshot 2023-06-26 at 3 48 02 PM
Screenshot 2023-06-26 at 3 48 14 PM
Screenshot 2023-06-26 at 3 48 22 PM

### Deleting Metric
Screenshot 2023-06-26 at 3 51 13 PM
Screenshot 2023-06-26 at 3 51 17 PM

### Documentation
Screenshot 2023-06-26 at 3 50 58 PM
Screenshot 2023-06-26 at 3 51 01 PM

Screenshot 2023-06-26 at 3 51 03 PM

### CODE REVIEW

**Comment by bttf** [comment]
> @bttf Just wanted to let you know we used this branch to automatically import all of our metrics. Haven't had a chance to test edge cases around modifying metrics that are assigned to experiments or not, but it looks good to us so far. Happy to iterate on it if we find any issues down the line.

That's great to hear, thank you for testing it out. Hoping to get it reviewed and merged soon. Let us know if you have any feedback in the meantime

File: packages/back-end/src/services/experiments.ts [hunk 1]
```diff
@@ -933,15 +933,11 @@ export function toSnapshotApiInterface(
...
- */
-export function postMetricApiPayloadIsValid(
-  payload: z.infer<typeof postMetricValidator.bodySchema>,
-  datasource: Pick<DataSourceInterface, "type">
-): { valid: true } | { valid: false; error: string } {
+const postOrPutMetricPayloadIsValid = (
+  payload:
+    | z.infer<typeof postMetricValidator.bodySchema>
+    | z.infer<typeof putMetricValidator.bodySchema>
+): { valid: true } | { valid: false; error: string } => {
```
Comment (tinahollygb) [inline packages/back-end/src/services/experiments.ts:940]: This doesn't seem like the right abstraction. Any changes you make here can impact the post endpoint. I think it makes sense to have a separate validator and not create this abstraction in an attempt to DRY up the code.

I see the `postMetricApiPayloadIsValid` is defined below—did you attempt to do a 2-in-1 validator here, realize it didn't work, and forget to restore the name? Or is the other one unused?

File: packages/back-end/src/services/experiments.ts [hunk 2]
```diff
@@ -1051,12 +1047,6 @@ export function postMetricApiPayloadIsValid(
         valid: false,
         error: "Binomial metrics cannot have an eventValue",
       };
-
-    if (datasource.type !== "mixpanel")
```
Comment (tinahollygb) [inline packages/back-end/src/services/experiments.ts:1055]: Do we enforce that datasource.type cannot be mixpanel?

File: packages/back-end/src/services/experiments.ts [hunk 3]
```diff
@@ -1075,6 +1065,40 @@ export function postMetricApiPayloadIsValid(
...
+    };
+  }
+
+  return postOrPutMetricPayloadIsValid(payload);
+}
+
+export function putMetricApiPayloadIsValid(
+  payload: z.infer<typeof putMetricValidator.bodySchema>
+): { valid: true } | { valid: false; error: string } {
+  return postOrPutMetricPayloadIsValid(payload);
```
Comment (tinahollygb) [inline packages/back-end/src/services/experiments.ts:1101]: I see  what you're doing here re: the abstraction. I think copying what's needed is the best approach rather than trying to DRY it up, because now the post validation is segmented across 2 separate functions.

**Review by tinahollygb — CHANGES_REQUESTED** [review]
You're missing the necessary permission checks.

I also left some comments about creating the wrong abstraction and potentially introducing a regression to the post metric validation. As discussed in our style guide, the rule of 3 should apply here. The word `or` in the function name is evident that this function is the wrong abstraction.

**Comment by bttf** [comment]
@tinahollygb Removed the abstraction as you suggested ... Also added permission checks for the `POST`, `PUT` and `DELETE` endpoints for Metrics. Let me know if I missed anything ... Thanks

File: packages/back-end/src/api/metrics/postMetric.ts [hunk 4]
```diff
@@ -21,6 +21,8 @@ export const postMetric = createApiRequestHandler(postMetricValidator)(
       throw new Error(`Invalid data source: ${datasourceId}`);
     }
 
+    req.checkPermissions("createMetrics", req.body?.projects ?? "");
```
Comment (tinahollygb) [inline packages/back-end/src/api/metrics/postMetric.ts:24]: Thanks for adding this.

**Review by tinahollygb — APPROVED** [review]
I didn't look too closely at the logic in the put handler so you may want to get other feedback in that area specifically, but at a glance it looks good. 

My previous concerns were addressed, and delete looks good.

**Comment by lukesonnet** [comment]
Does remove the abstraction also necessitate duplicating all the YAML code in `packages/back-end/src/api/openapi/payload-schemas/*`?

Mostly just trying to understand what's going on here, as this will mean having to double the amount of API work for any future change to the metrics interface.

**Comment by bttf** [comment]
> One small thing we noticed is that the `dateUpdated` isn't actually updated on `PUT` of a metric change

Great catch - this seemed to be an issue app-wide; fixed and tested. Changes pushed

**Comment by bttf** [comment]
> Does remove the abstraction also necessitate duplicating all the YAML code in `packages/back-end/src/api/openapi/payload-schemas/*`?
> 
> Mostly just trying to understand what's going on here, as this will mean having to double the amount of API work for any future change to the metrics interface.

The payloads are similar but their validation rules are dissimilar - we allow the omission of most fields on the metric interface for the PUT endpoint. We don't enforce required fields except when necessary for deeply nested fields like `identifierTypeColumns` (which require a column name and data type to store the column). It's difficult to consolidate the YAML using references when required rules differ further down the tree. 

You're right that it's technically double the work should we decide to update the metric interface however the necessary changes would be surface level (regarding API interface) and the brunt of the work still is handled by the same backend method `updateMetric` which both endpoints share.
